src/storage/log/mod.rs

```diff
@@ -0,0 +1,2388 @@
+        len_buf.copy_from_slice(&(self.data.len() as u32).to_be_bytes());
+        writer.write_all(&len_buf)?; // Write length
+
+        // Loop to write key-value pairs to the writer
+        for (key, value) in &self.data {
+            write_field(key.as_bytes(), writer)?; // Write key
+            write_field(value, writer)?; // Write value
+        }
+
+        Ok(())
+    }
+
+    // Puts an integer value with a given key
+    pub(crate) fn put_uint(&mut self, key: &str, n: u64) {
+        let b = n.to_be_bytes(); // Convert integer to big-endian bytes
+        self.put(key, &b); // Call the generic put method
+    }
+
+    // Gets an integer value associated with a given key
+    pub(crate) fn get_uint(&self, key: &str) -> Result<u64> {
+        // // Use the generic get method to retrieve bytes
+        let value_bytes = self.get(key).ok_or(Error::IO(IOError::new(
+            io::ErrorKind::NotFound,
+            "Key not found",
+        )))?;
+
+        let bytes: &[u8] = value_bytes.as_slice();
+        let value_bytes = bytes.as_ref().try_into().map_err(|_| {
+            Error::IO(IOError::new(
+                io::ErrorKind::InvalidData,
+                "Failed to convert bytes to u64",
+            ))
+        })?;
+
+        let int_value = u64::from_be_bytes(value_bytes);
+        Ok(int_value)
+    }
+
+    pub(crate) fn put_bool(&mut self, key: &str, b: bool) {
+        let value = if b { 1 } else { 0 };
+        self.put_uint(key, value);
+    }
+
+    pub(crate) fn get_bool(&self, key: &str) -> Result<bool> {
+        let value = self.get_uint(key)?;
+        Ok(value == 1)
+    }
+
+    // Generic method to put a key-value pair into the data HashMap
+    pub(crate) fn put(&mut self, key: &str, value: &[u8]) {
+        self.data.insert(key.to_string(), value.to_vec());
+    }
+
+    // Generic method to get the value associated with a given key
+    pub(crate) fn get(&self, key: &str) -> Option<&Vec<u8>> {
+        self.data.get(key)
+    }
+}
+
+/// Enum representing different types of records in a write-ahead log.
+///
+/// This enumeration defines different types of records that can be used in a write-ahead log.
+/// Each record type indicates a particular state of a record in the log. The enum variants
+/// represent various stages of a record, including full records, fragments, and special cases.
+///
+/// # Variants
+///
+/// - `Empty`: Indicates that the rest of the block is empty.
+/// - `Full`: Represents a full record.
+/// - `First`: Denotes the first fragment of a record.
+/// - `Middle`: Denotes middle fragments of a record.
+/// - `Last`: Denotes the final fragment of a record.
+#[derive(PartialEq)]
+enum RecordType {
+    Empty = 0,  // Rest of block is empty.
+    Full = 1,   // Full record.
+    First = 2,  // First fragment of a record.
+    Middle = 3, // Middle fragments of a record.
+    Last = 4,   // Final fragment of a record.
+}
+
+impl RecordType {
+    fn from_u8(value: u8) -> Result<Self> {
```

Comment:
Have you considered implementing `TryFrom<u8> for RecordType` instead?

Reply:
Thanks for suggesting this, I haven't, just looking into it.

Reply:
I've sort of wrapped the Result here like this `pub type Result<T> = std::result::Result<T, Error>;`, and using tryfrom requires conversion between the error types

Reply:
Isn't `Error` public in this case? You should be fine if you use the same error type in your TryFrom implementation. It's fine though. I'm not hung up on this. Using TryFrom would publicly expose this implementation anyway if RecordType is public.